api-examples/transformer_utils.py

```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
+        :param unif: An optional initialization value which can set the linear weights.  If given, biases will init to 0
+        :param initializer: An initialization scheme by string name: `ortho`, `kaiming` or `he`, `xavier` or `glorot`
+        """
+        super().__init__()
+        if insz == outsz:
+            self.layer = SkipConnection(insz, activation)
+        else:
+            self.layer = Dense(insz, outsz, activation)
+
+        self.output_dim = outsz
+        self.ln = nn.LayerNorm(outsz, eps=1e-6)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        """Run a linear projection over the input, followed by an optional activation given by constructor
+
+        :param input: the input tensor
+        :return: the transformed output
+        """
+        return self.ln(self.layer(input))
+
+
+class DenseLNStack(nn.Module):
+    """A stack of one or more hidden layers
+    """
+
+    def __init__(
+            self,
+            insz: int,
+            hsz: Union[int, List[int]],
+            activation: str = "gelu",
+            pdrop_value: float = 0.5,
+            init=None,
+            skip_connect=True,
+            **kwargs,
+    ):
+        """Stack 1 or more hidden layers, optionally (forming an MLP)
+
+        :param insz: The number of input units
+        :param hsz: The number of hidden units
+        :param activation: The name of the activation function to use
+        :param pdrop_value: The dropout probability
+        :param init: The initializer
+        :param skip_connect: whether use an overall skip connection
+        """
+        super().__init__()
+        hszs = listify(hsz)
+        self.output_dim = hsz[-1]
+        current = insz
+        layer_stack = []
+        for hsz in hszs[:-1]:
+            layer_stack.append(WithDropout(DenseLN(current, hsz, activation=activation), pdrop_value))
+            current = hsz
+        self.layer_stack = nn.Sequential(*layer_stack)
+        self.final = Dense(current, self.output_dim)
+        self.ln1 = nn.LayerNorm(insz, eps=1e-6)
+        self.ln2 = nn.LayerNorm(self.output_dim, eps=1e-6)
+        self.skip_connect = skip_connect
+        if self.skip_connect:
+            self.proj = Dense(insz, self.output_dim) if insz != self.output_dim else nn.Identity()
+        self.requires_length = False
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Stack 1 or more hidden layers, optionally (forming an MLP)
+
+        :param inputs: The fixed representation of the model
+
+        :Keyword Arguments:
+        * *hsz* -- (``int``) The number of hidden units (defaults to `100`)
+
+        :return: The final layer
+        """
+        x = self.ln1(inputs)
+        for layer in self.layer_stack:
+            x = layer(x)
+        x = self.final(x)
+        if self.skip_connect:
+            x = x + self.proj(inputs)
+        return self.ln2(x)
```

Comment:
ln2 shouldn't be there

Reply:
actually dont think 189-191 belongs in DenseStack.  Also DenseStack should be simplified to just call the sequential's forward

Reply:
you can get the same functionality in the dual encoder by having a `nn.Sequential(DenseStack, ResidualOutputLayer)`